### PR TITLE
Support for natural light for Kobo Aura One

### PR DESCRIPTION
**Pull request by dengste**
This adds support for accessing the Kobo Aura One frontlight over the sysfs interface, which should fix the low brightness issue (see #2321)  and adding support for natural light (see #2285). Currently, the only way to change the natural light is through to Kobolight plugin's gesture controller, but I plan on adding a widget for this as well.

### CODE REVIEW

File: plugins/kobolight.koplugin/main.lua [hunk 1]
```diff
@@ -153,19 +199,47 @@ function KoboLight:onSwipe(_, ges)
...
+
 function KoboLight:addToMainMenu(menu_items)
     menu_items.frontlight_gesture_controller = {
         text = _("Frontlight gesture controller"),
         callback = function()
+            local image_name
+            local nl_text = ""
+            if with_natural_light then
+                image_name = "/demo_ka1.png"
+                nl_text = "\n- Change frontlight warmth by swiping up or down on the right of the screen."
```
Comment (Frenzie) [inline plugins/kobolight.koplugin/main.lua]: Needs `_()` for internationalization.

File: frontend/device/kobo/sysfs_light.lua [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+end
+
+function KoboSysfsLight:setWarmth(warmth)
+    assert(warmth >= 0 and warmth <= 100,
+           "Wrong warmth value given!")
+    self:setNaturalBrightness(self.current_brightness, warmth)
+end
+
+function KoboSysfsLight:setNaturalBrightness(brightness, warmth)
+    -- On Nickel, the values for white/red/greem are roughly linearly dependend
```
Comment (Frenzie) [inline frontend/device/kobo/sysfs_light.lua]: With the very high quality of this PR, the typos in this comment stick out to me a bit. 😊 

(green, dependent)

File: frontend/device/kobo/nickel_conf.lua [hunk 3]
```diff
@@ -143,4 +155,12 @@ function NickelConf.frontLightState.set(new_state)
                                        true)
 end
 
+function NickelConf.colorSetting.set(new_color)
+    assert(new_color >= 1500 and new_color <= 6400,
```
Comment (Frenzie) [inline frontend/device/kobo/nickel_conf.lua]: Note that asserts and errors generally speaking mean insta-crash. As such we want to restrict them to debug-mode only, see

https://github.com/koreader/koreader/blob/775b7a364f4b0a1b82fc236d4aaa6513922f9904/frontend/ui/uimanager.lua#L298-L302

You can also do things with if dbg.is_on but that doesn't seem appropriate here (nor does pcall).

File: frontend/device/kobo/powerd.lua [hunk 4]
```diff
@@ -32,6 +35,14 @@ function KoboPowerD:_syncKoboLightOnStart()
         elseif kobo_light_on_start == -2 then -- get values from NickelConf
             new_intensity = NickelConf.frontLightLevel.get()
             is_frontlight_on = NickelConf.frontLightState:get()
+            if self.fl_warmth ~= nil then
+                local new_color = NickelConf.colorSetting.get()
+                if new_color ~= nil then
+                    -- ColorSetting is in [1500,6400], with '1500'
+                    -- being maximum warmth, so normalize this to [0,100]
+                    new_warmth = (100 - math.floor((new_color - 1500) / 49))
```
Comment (Frenzie) [inline frontend/device/kobo/powerd.lua]: Losing over a magnitude of precision is no concern?

**Comment by dengste** [comment]
I've added fixups regarding the asserts, usage of '_' and the typos; please squash them if they are OK. The luacheck thingy on CircleCI does a segfault, but I can't see how my changes could relate to that. Regarding the loss of precision of the NickelConf 'ColorSetting' value: It's not like you can actually set 4900 different warmth values through the slider in Nickel's GUI, the slider is not nearly precise enough. So I think mapping this value to full percentages is OK.

**Comment by dengste** [comment]
Oh, an feel free to merge the two math.pow calls into one if you wish (see [here](https://github.com/koreader/koreader/issues/2321#issuecomment-362533102)). I've left them separate so that it becomes clear that those exponents may be changed separately. I don't think it makes a noticeable difference in performance, but of course I wouldn't mind doing this optimization.

**Comment by Frenzie** [comment]
> It's not like you can actually set 4900 different warmth values through the slider in Nickel's GUI

We're not bound by Nickel, of course. ;-)

> I don't think it makes a noticeable difference in performance, but of course I wouldn't mind doing this optimization.

I'd say that's just for people, certainly nothing noticeable.

**Comment by poire-z** [comment]
You could say a word about this being available in current nightlies, in the 2 original issues you referenced above, so that all the people that contributed there get a mail and can know it's done, and they can test and enjoy it :)

**Comment by Frenzie** [comment]
Don't you also get a notification about mentions in other issues?

**Comment by poire-z** [comment]
I don't think we do, for mentions.
But we get one when the issue is closed, even when automatically closed when a PR is merged.

**Comment by bendtherules** [comment]
Just a stupid question, how do I try it out right now from nightly build?
Waiting for a long time, really excited to see this land.

**Comment by Frenzie** [comment]
It's slightly hidden under More plugins → Fronlight gesture controller.

screenshot_2018-02-11_07-12-13
screenshot_2018-02-11_07-12-19
screenshot_2018-02-11_07-12-27

**Pull request merged**